processor/src/main/java/org/mapstruct/ap/internal/util/workarounds/EclipseAsMemberOfWorkaround.java

```diff
@@ -0,0 +1,196 @@
+ *  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+package org.mapstruct.ap.internal.util.workarounds;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.Comparator;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+
+import javax.annotation.processing.ProcessingEnvironment;
+import javax.lang.model.element.Element;
+import javax.lang.model.type.DeclaredType;
+import javax.lang.model.type.TypeMirror;
+import javax.lang.model.util.Types;
+
+import org.eclipse.jdt.core.compiler.CharOperation;
+import org.eclipse.jdt.internal.compiler.apt.dispatch.BaseProcessingEnvImpl;
+import org.eclipse.jdt.internal.compiler.apt.model.ElementImpl;
+import org.eclipse.jdt.internal.compiler.lookup.MethodBinding;
+import org.eclipse.jdt.internal.compiler.lookup.ReferenceBinding;
+import org.eclipse.jdt.internal.compiler.lookup.TypeBinding;
+
+/**
+ * Contains the workaround for {@link Types#asMemberOf(DeclaredType, Element)} using Eclipse implementation types.
+ * <p>
+ * <strong>This class may only be loaded when running within Eclipse</strong>
+ *
+ * @author Andreas Gudian
+ */
+final class EclipseAsMemberOfWorkaround {
+    private EclipseAsMemberOfWorkaround() {
+    }
+
+    /**
+     * Eclipse-specific implementation of {@link Types#asMemberOf(DeclaredType, Element)}.
+     * <p>
+     * Returns {@code null} if the implementation could not determine the result.
+     */
+    static TypeMirror asMemberOf(ProcessingEnvironment environment, DeclaredType containing,
+                                 Element element) {
+
+        ElementImpl elementImpl = tryCast( element, ElementImpl.class );
+        BaseProcessingEnvImpl env = tryCast( environment, BaseProcessingEnvImpl.class );
+
+        if ( elementImpl == null || env == null ) {
+            return null;
+        }
+
+        ReferenceBinding referenceBinding =
+            (ReferenceBinding) ( (ElementImpl) environment.getTypeUtils().asElement( containing ) )._binding;
+
+        MethodBinding methodBinding = (MethodBinding) elementImpl._binding;
+
+        // matches in super-classes have priority
+        MethodBinding inSuperclassHiearchy = findInSuperclassHierarchy( methodBinding, referenceBinding );
+
+        if ( inSuperclassHiearchy != null ) {
+            return env.getFactory().newTypeMirror( inSuperclassHiearchy );
+        }
+
+        // if nothing was found, traverse the interfaces and collect all candidate methods that match
+        List<MethodBinding> candidatesFromInterfaces = new ArrayList<MethodBinding>();
+
+        collectFromInterfaces(
```

Comment:
It could also be that `containing` is an abstract class that we start searching from. Only if that abstract class already contains an implementation of the method we're looking for, we can stop searching (as it can't be any _less_ specific than what's declared in the interfaces).